feat(token-selector): show default tokens for connected wallets with no balance

jjramirezn opened the pull request:
This PR adds two things:

- In the request pay flow, we indicate that wallet should be connected
- In all flows, if there is a connected wallet that has no balances, show top 6 tokens when the token selector is opened

jjramirezn committed:
fix(request): require wallet connection on pay

jjramirezn committed:
feat: add default tokens for connected wallets with no balance

jjramirezn committed:
refactor: fallback on chainId when populating default tokens

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

Hugo0 commented on src/components/Global/TokenSelector/TokenSelector.tsx:
```diff
@@ -94,8 +95,23 @@ const TokenSelector = ({ classNameButton, shouldBeConnected = true, onReset }: _
+                amount: 0,
+                currency: '',
+                value: '',
+            }))
```
nit: I think ideally we'd have a comment here in the code explaining why top 6 is the choice (and not top 200 or why we dont need to rank them)

jjramirezn replied:
changed quite a bit, added comment on filters

Hugo0 commented on src/components/Global/TokenSelector/TokenSelector.tsx:
```diff
@@ -94,8 +95,23 @@ const TokenSelector = ({ classNameButton, shouldBeConnected = true, onReset }: _
+
         return balancesToDisplay
-    }, [balances, safeInfo])
+    }, [balances, safeInfo, selectedChainID])
```
ncie

Hugo0 reviewed (approved):
nicee

Hugo0 reviewed (changes requested):
Actually, reviewing the behaviour in the [vercel deployment](https://peanut-ui-git-feat-selector-view-disconnected-squirrellabs.vercel.app/request/create), 2 issues:

1. when initially loading the page and clicking on tokenSelector quickly, it goes to the "explore & buy more tokens" view, and only after waiting a second or two it goes to the first page

2. I think it shouldn't go to the first page, i think these 6 tokens should be on the search page (not the initial page):
![image](https://github.com/user-attachments/assets/7009ee6d-e4e2-4e21-a2ae-05757d7b6fc7)
it's more natural for the user to view that view, and if they want a different token then they can enter some input in the search bar

Like this:
![image](https://github.com/user-attachments/assets/988a0633-9611-4072-896f-1edcd6524fe9)

jjramirezn committed:
refactor(token-selector): remove hardcoded arbitrum logo

Checked that the logo is retreived fine for Arbitrum One chain, hardcode
not needed anymore.

jjramirezn committed:
refactor(token-selector): abstract token list

jjramirezn committed:
feat(token-selector): new token selector

Now we show the balances and the search bar in the same view, saving one
click. We also make the balances searchable together with the tokens
that have no balances.

jjramirezn committed:
fix(modal): positioning for small screens

it was sticking to the bottom of the screen, now it's centered

jjramirezn commented on src/components/Global/TokenSelector/TokenSelector.tsx:
```diff
@@ -94,8 +205,23 @@ const TokenSelector = ({ classNameButton, shouldBeConnected = true, onReset }: _
+
         return balancesToDisplay
-    }, [balances, safeInfo])
+    }, [balances, safeInfo, selectedChainID])
```
you are right and I added it, thanks

jjramirezn commented on src/components/Global/TokenSelector/TokenSelector.tsx:
```diff
@@ -18,90 +19,203 @@ import { useWalletType } from '@/hooks/useWalletType'
+                                                className="absolute left-0 top-0 h-6 w-6"
+                                                alt="logo"
+                                                onError={(e) => {
+                                                    console.log(e)
```
i did not introduce this log, but I agree lets remove it

jjramirezn committed:
refactor: remove unnecessary log

Co-authored-by: coderabbitai[bot] <136622811+coderabbitai[bot]@users.noreply.github.com>

jjramirezn commented on src/components/Global/TokenSelector/TokenSelector.tsx:
```diff
@@ -18,90 +19,203 @@ import { useWalletType } from '@/hooks/useWalletType'
+                    balances.map((balance, idx) => (
+                        <tr
+                            key={idx}
+                            className={`h-14 cursor-pointer gap-0 transition-colors hover:bg-n-3/10 ${utils.areTokenAddressesEqual(balance.address, selectedTokenAddress) && balance.chainId === selectedChainID && `bg-n-3/10`}`}
```
what?

jjramirezn committed:
refactor(token-selector): reduce filter clean timeout

jjramirezn committed:
refactor(token-selector): reset selected balance when chain changes

jjramirezn commented on src/components/Global/TokenSelector/TokenSelector.tsx:
```diff
@@ -18,90 +19,203 @@ import { useWalletType } from '@/hooks/useWalletType'
             setFilterValue('')
-            setShowFallback(false)
         }, 500)
         setVisible(false)
```
changed to 200ms to be in line with the fade-out of the modal

jjramirezn commented on src/components/Global/TokenSelector/TokenSelector.tsx:
```diff
@@ -112,20 +226,22 @@ const TokenSelector = ({ classNameButton, shouldBeConnected = true, onReset }: _
+            )
         }
-    }, [displayedToken])
+    }, [_balancesToDisplay, selectedTokenAddress, selectedChainID])
```
applied

jjramirezn commented on src/components/Global/TokenSelector/TokenSelector.tsx:
```diff
@@ -18,92 +19,204 @@ import { useWalletType } from '@/hooks/useWalletType'
-        }, 500)
-        setVisible(false)
+        }, 200) // the modal takes 200ms to close (duration-200 on leave)
     }
```
it's ok

jjramirezn commented on src/components/Global/TokenSelector/TokenSelector.tsx:
```diff
@@ -18,92 +19,204 @@ import { useWalletType } from '@/hooks/useWalletType'
+        if (filterValue.length === 0 && !hasUserChangedChain) return _balancesToDisplay
 
-        if (safeInfo && walletType === 'blockscout') {
-            balancesToDisplay = balances.filter((balance) => balance.chainId.toString() === safeInfo.chainId.toString())
-        }
+        const byChainAndText = ({ name, symbol, chainId }: IUserBalance): boolean =>
+            // if the user has changed chains, only show tokens on the current chain
+            // even if they have balances on other chains
+            (!hasUserChangedChain || selectedChainID === chainId) &&
+            // if the user has typed something, only show tokens that match
+            (filterValue.length === 0 ||
+                name.toLowerCase().includes(filterValue.toLowerCase()) ||
+                symbol.toLowerCase().includes(filterValue.toLowerCase()))
 
-        return balancesToDisplay
-    }, [balances, safeInfo])
+        return _balancesToDisplay.filter(byChainAndText)
+    }, [_balancesToDisplay, filterValue, selectedChainID, hasUserChangedChain])
```
not an expensive calculation to merit useCallback

Hugo0 commented on src/components/Global/TokenSelector/Components/AdvancedButton.tsx:
```diff
@@ -142,23 +118,20 @@ export const AdvancedTokenSelectorButton = ({
                             aria-label="Reset token selection"
                             className="block"
                             onClick={(e) => {
-                                e.stopPropagation()  //don't open modal
```
praise: lol, nice. consistent formatting ftw

jjramirezn committed:
refactor(token-selector): show all balances after modal is reopened

jjramirezn commented on src/components/Global/TokenSelector/TokenSelector.tsx:
```diff
@@ -18,92 +19,205 @@ import { useWalletType } from '@/hooks/useWalletType'
+        if (filterValue.length === 0 && !hasUserChangedChain) return _balancesToDisplay
 
-        if (safeInfo && walletType === 'blockscout') {
-            balancesToDisplay = balances.filter((balance) => balance.chainId.toString() === safeInfo.chainId.toString())
-        }
+        const byChainAndText = ({ name, symbol, chainId }: IUserBalance): boolean =>
+            // if the user has changed chains, only show tokens on the current chain
+            // even if they have balances on other chains
+            (!hasUserChangedChain || selectedChainID === chainId) &&
+            // if the user has typed something, only show tokens that match
+            (filterValue.length === 0 ||
+                name.toLowerCase().includes(filterValue.toLowerCase()) ||
+                symbol.toLowerCase().includes(filterValue.toLowerCase()))
 
-        return balancesToDisplay
-    }, [balances, safeInfo])
+        return _balancesToDisplay.filter(byChainAndText)
+    }, [_balancesToDisplay, filterValue, selectedChainID, hasUserChangedChain])
```
non expensive function

Hugo0 commented on src/components/Global/TokenSelector/TokenSelector.tsx:
```diff
@@ -18,92 +19,205 @@ import { useWalletType } from '@/hooks/useWalletType'
+    const [selectedBalance, setSelectedBalance] = useState<IUserBalance | undefined>(undefined)
+    const [hasUserChangedChain, setUserChangedChain] = useState(false)
     const focusButtonRef = useRef<HTMLButtonElement>(null)
-    /**
-     * If the user is not connected, we show a fallback screen to search for tokens directly,
-     * instead of showing the balances.
-     */
-    const [showFallback, setShowFallback] = useState(!shouldBeConnected)
```
praise: love the new behaviour

Hugo0 commented on src/components/Global/TokenSelector/TokenSelector.tsx:
```diff
@@ -134,33 +252,28 @@ const TokenSelector = ({ classNameButton, shouldBeConnected = true, onReset }: _
-                tokenSymbol={displayedToken?.symbol ?? ''}
-                tokenBalance={displayedTokenBalance?.amount ?? 0}
-                chainIconUri={(IconPlaceholderChecker(selectedChainID) as string) ?? displayedChain?.icon.url}
+                tokenLogoUri={selectedBalance?.logoURI ?? ''}
```
question: what if the logoUri is empty? Then we get a broken looking icon

Can we use the fallback icon we have alrdy, like in this example:
![image](https://github.com/user-attachments/assets/706b9d81-f468-4810-bcf5-b0e1367b54fa)

jjramirezn replied:
good idea, added

Hugo0 reviewed (approved):
Generally looks good, didn't go to deep in the review - lfg

jjramirezn committed:
refactor(token-selector): add placeholder icons fror button
